pkg/tarbackfill/tarbackfill.go

```diff
@@ -0,0 +1,159 @@
+package tarbackfill
+
+import (
+	"archive/tar"
+	"io"
+	"path"
+	"strings"
+)
+
+// Reader wraps a tar.Reader so that if an item which would be read from it is
+// in a directory which is not included in the archive, a specified Backfiller
+// interface's Backfill() method will be called to supply a tar.Header which
+// will be inserted into the stream just ahead of that item.
+type Reader struct {
+	*tar.Reader
+	backfiller      Backfiller
+	seen            map[string]struct{}
+	queue           []*tar.Header
+	currentIsQueued bool
+	err             error
+}
+
+// Backfiller is a wrapper for Backfill, which can supply headers to insert
+// into an archive which is on its way to being extracted.
+type Backfiller interface {
+	// Backfill either returns an entry for the passed-in path, nil if
+	// no entry should be added to the stream, or an error if something
+	// unexpected happened.
+	Backfill(string) (*tar.Header, error)
+}
+
+// NewReaderWithBackfiller creates a new Reader reading from r, asking the
+// passed-in Backfiller for information about parent directories which it
+// hasn't seen yet.
+func NewReaderWithBackfiller(r *tar.Reader, backfiller Backfiller) *Reader {
+	reader := &Reader{
+		Reader:     r,
+		backfiller: backfiller,
+		seen:       make(map[string]struct{}),
+	}
+	return reader
+}
+
+// Next returns either the next item from the archive we're filtering, or a
+// synthesized entry for a directory that arguably should have been in that
+// archive.
+func (r *Reader) Next() (*tar.Header, error) {
+	// Drain the queue first.
+	if len(r.queue) > 0 {
+		next, queue := r.queue[0], r.queue[1:]
+		r.queue = queue
+		r.currentIsQueued = len(r.queue) > 0
```

Comment:
Shouldn’t this actually be true for this just-dequeued entry as well, i.e. just `true` unconditionally on this path?